src/models/generalized_mixture_reach_surface.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2021 The Private Cardinality Estimation Framework Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Class for modeling Pairwise union reach surface."""
+
+import copy
+from typing import List
+import numpy as np
+from cvxopt import matrix
+from cvxopt import solvers
+from typing import Iterable
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.models.reach_surface import ReachSurface
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+
+solvers.options["show_progress"] = False
+
+
+class GeneralizedMixtureReachSurface(ReachSurface):
+    """Models reach with the generalized mixture overlap model.
+
+    GeneralizedMixtureReachSurface models the combined reach with the formula
+
+
+    N * sum_{j=1..c} [1 - prod_{i=1..p} (1-a_i_j r_i / N)] +epsilon
+
+    where :
+          all a_i_j >= 0
+          sum_{j=1..c} a_i_j=1 for all i.
+
+    """
+
+    def __init__(
+        self,
+        reach_curves: Iterable[ReachCurve],
+        reach_points: Iterable[ReachPoint],
+        num_clusters: int,
+        N: int,
+    ):
+        """Constructor for GeneralizedMixtureReachSurface.
+
+        Args:
+          reach_curves: A list of ReachCurves to be used in model fitting and
+            reach prediction.
+          reach_points: A list of ReachPoints to which the model is to be fit.
+            This list is of arbitrary length, but it should contain a minimum of
+            p(p-1)/2 points, where p is the number of publishers. Otherwise,
```

Comment:
+1 (guess it was a typo).

@matthewclegg  For your question "What will happen if fewer than c*(p-1) points are given", there will be error since this solution uses a quadratic programming and the QP no longer has a unique solution when #training points < c(p - 1)

Reply:
Thanks Matthew, updated it.